src/sync.ts

```diff
@@ -209,6 +228,15 @@ export class SyncApi {
             RoomEvent.TimelineReset,
         ]);
         this.registerStateListeners(room);
+        // Register listeners again after the state reference changes
+        room.on(RoomEvent.CurrentStateUpdated, (targetRoom, previousCurrentState) => {
+            if (targetRoom !== room) {
+                return;
+            }
+
+            this.deregisterStateListeners(previousCurrentState);
+            this.registerStateListeners(room);
+        });
```

Comment:
Calling out this new behavior.

This is necessary because `room.fixUpLegacyTimelineFields()` (and `room.resetLiveTimeline()` which calls that function) will change the `room.currentState` reference and we will fail listen to any new state events.

The one place we deregistered and registered before was for a `limited` `/sync` response but that was only possible because that is handled in `sync.ts` (see below in this file https://github.com/matrix-org/matrix-js-sdk/pull/2299#discussion_r882335575)

Now we're also calling `fixUpLegacyTimelineFields` inside of `room.ts` so we need to make sure any consumers of the room know about the reference change.

Reply:
This feels like a breaking change, all downstream clients will need updating if any hold references to the currentState

Reply:
@t3chguy All existing functionality still works.

Clients who choose to implement timeline refreshing will also need to take `room.currentState` changing though.